### DISCUSSION TITLE
Are multiple windows of an app supported?

### DISCUSSION THREAD

**Discussion by tukusejssirs** [Nama/sway-workspaces#6] [Q&A] [answered]
I have read [this comment](https://github.com/swaywm/sway/issues/1005#issuecomment-1278150088) and [your response](https://github.com/swaywm/sway/issues/1005#issuecomment-1278330219). I have searched PRs, but I could not find a PR by [@roselandgoose](https://github.com/roselandgoose) nor a PR related to multiple windows of an app. It doesn’t seem to work right now (tested `v1.0` from GitHub Releases).

Although [@roselandgoose](https://github.com/roselandgoose) claimed she added _support multiple windows of the same program_. Her version of [`sway_workspaces.py`](https://git.sr.ht/~roselandgoose/Sway_Layout_Restore/blob/main/sway_workspaces.py) still contains `Multiple windows of an application/class aren't handled. No idea how to do so.` comment. I haven’t tested her version of the script.

I simply want my windows (even those of a same program) open in a specific order right after Sway startup. I presume I need to start the programs first, then run `sway_workspaces.py` and it will re-order the windows based on the saved configuration, right?

After saving the current config, reordering the windows (both windows of an app and those of different apps), I tried to restore the window order using `load` command, but the windows order didn’t change. Tried to re-order both XWayland and Wayland windows (like JetBrains Webstorm [X], Sublime Text [W], Alacritty [W], Insomnia [X]), but it does doesn’t do anything but blinks the windows (actually entire output on each monitor) and outputs `Loaded Workspace Setup` notification.

I use `sway version 1.9-dev-ab9b164e (Jan 31 2023, branch 'master')` on Arch Linux if that matters.

**Comment by Nama** [comment, marked as answer]
I wanted rose to do a PR, but they didn't. So I made 7580bc1f9151d2980c0ccd635b1822c1824675cd mostly of roses code.

> Her version of [sway_workspaces.py](https://git.sr.ht/~roselandgoose/Sway_Layout_Restore/blob/main/sway_workspaces.py) still contains `Multiple windows of an application/class aren't handled. No idea how to do so.` comment.

That's just because they didn't care of the Readme. And I just noticed I didn't write about that in the README.md, too D:
But here #3

> I presume I need to start the programs first, then run sway_workspaces.py and it will re-order the windows based on the saved configuration, right?

Exactly. just re-run `sway_workspaces load <profile>` every time you want to re-order your windows.

> I haven’t tested her version of the script.

Please do, just to see if any other commits here are breaking that feature. Especially this one 53bff61a2c19087958aec68474fd10084539a046

> It doesn’t seem to work right now (tested v1.0 from GitHub Releases).

For me, just two Firefox windows, it mostly worked, if I have the same tabs in the foreground, which I had while saving the profile. If the title of the window doesn't match, it can't be put in the correct workspace.

But I guess, 53bff61a2c19087958aec68474fd10084539a046 to don't touch the windows which are on the "correct" workspace, breaks this. Since the title isn't checked.

> After saving the current config, reordering the windows 

You need to save the profile after you reordered your windows manually. Or do you mean "change, to test if it works"?.

**Reply by tukusejssirs** [reply]
> [[src](https://github.com/Nama/sway-workspaces/issues/3#issuecomment-1280110354)]
>
> Window titles are compared. That means, that if you have another website open in your browser, it wont be possible to match. But then it still moves them separately to the correct workspaces. Might just be the wrong windows^^

`Window titles are compared`: why didn’t use `.nodes.nodes[].node[].id`s? Like:

```bash
swaymsg -t get_tree | jq '.nodes[].nodes[].nodes[].nodes[].id'
```

Outupts unique nodes for all my Firefox and Sublime Text windows. IMHO that way we could support windows of an app regardless of the window titles. Or do I miss something? :thinking:

---

I assign some apps to specific Sway workspaces; cannot it interfere with `sway_workspaces`?

---

> > I haven’t tested her version of the script.
>
> Please do, just to see if any other commits here are breaking that feature. Especially this one [53bff61](https://github.com/Nama/sway-workspaces/commit/53bff61a2c19087958aec68474fd10084539a046)

I might do this, I just need to find some more time.

> > It doesn’t seem to work right now (tested v1.0 from GitHub Releases).
>
> For me, just two Firefox windows, it mostly worked, if I have the same tabs in the foreground, which I had while saving the profile. If the title of the window doesn't match, it can't be put in the correct workspace.

I use Firefox Nightly, but that shouldn’t actually matter. I run Firefox in Wayland. I assigned it to a specific workspace. If I move a Firefox window to a different workspace, load the config, the window is moved to the correct workspace, but not to the place where it was when I saved the config. For example, I have four Firefox windows open, I move the first one to a different workspace, load the config, the moved window is moved after the currectly active window on that workspace, not in the place where it was open (e.g. active window is the third window [window order: 2, 3, 4]; new order: 2, 3, 1, 4). I want the saved order of the windows to be preserved.

> But I guess, [53bff61](https://github.com/Nama/sway-workspaces/commit/53bff61a2c19087958aec68474fd10084539a046) to don't touch the windows which are on the "correct" workspace, breaks this. Since the title isn't checked.

Yes, it might be an issue, as the order of the windows is currently not checked (was it before 53bff61? I have no idea). Currently, even if there is only a single window of an app, although it is moved to the right workspace, the order of the windows (both of an app and of different apps) is not preserved.

IMHO it would be nice to leave windows intact _only if_ a window is on the correct workspace AND if it is in the correct order.

> > After saving the current config, reordering the windows
>
> You need to save the profile after you reordered your windows manually. Or do you mean "change, to test if it works"?.

What I meant is as follows:

1. `sway-workspaces save config`.
2. Manually reorder windows (just for the test purposes).
3. `sway-workspaces load config`.
4. The order of all windows on every workspace to be the same as when I run `sway-workspaces save config`, however, it is not the same.

**Reply by Nama** [reply]
> `Window titles are compared`: why didn’t use .nodes.nodes[].node[].ids? Like:

~~That doesn't even output anything for me (using sway 1.8). I also just noticed, that `swaymsg -t get_tree` isn't outputting a json anymore.~~ Needs to be piped.
This works: `swaymsg -t get_tree | jq '.nodes[].nodes[].nodes[].id'`
~~But I know what you mean: `#12: con "Element [1] |`
13 would be the ID of this node, but~~ that ID is not persistent. After a reboot, you will get new IDs.

> I assign some apps to specific Sway workspaces; cannot it interfere with `sway_workspaces`?

What do you mean? How do you assign apps to workspaces?

> I want the saved order of the windows to be preserved.

> Currently, even if there is only a single window of an app, although it is moved to the right workspace, the order of the windows (both of an app and of different apps) is not preserved.

This was never implemented. But I thought about it and might do it at a later time.

**Reply by tukusejssirs** [reply]
> > `Window titles are compared`: why didn’t use .nodes.nodes[].node[].ids? Like:
> 
> That doesn't even output anything for me (using sway 1.8). ~I also just noticed, that `swaymsg -t get_tree` isn't outputting a json anymore.~ Needs to be piped. But I know what you mean: `#12: con "Element [1] |` 13 would be the ID of this node, but that ID is not persistent. After a reboot, you will get new IDs.

Yeah, I believe in Sway starting with `v1.8`, `swaymsg` (`get_tree` and `get_outputs` at least) does not output JSON _when the output is not piped into another program_. When I use `swaymsg -t get_tree | less`, I get a JSON as expected. Now, if you don’t want to pipe the content, you can use `-r` option (i.e. `swaymsg -rt get_tree`; [src](https://man.archlinux.org/man/swaymsg.1.en)).

> that ID is not persistent

I get it. You have no way to identify a particular window of an app unless the window title does not change. Without that, I see no way to make it possible to make the order of windows of an app persistent across reboots, unfortunately.

> > I assign some apps to specific Sway workspaces; cannot it interfere with `sway_workspaces`?
> 
> What do you mean? How do you assign apps to workspaces?

In Sway config:

```bash
# Assign an app that has `app_id` to a workspace (prefered)
assign [app_id="Firefox"] workspace 3
# Assign an app that has `class` to a workspace (prefered unless the app has `app_id`)
assign [class="VSCodium"] workspace 2

# Move an app that has neither `app_id` nor `class` to a workspace
# Actually, this moves all windows that ends with `.*Firefox Nightly` regex.
for_window [title=".*Firefox Nightly"] move window to workspace 3
```

> > I want the saved order of the windows to be preserved.
> 
> > Currently, even if there is only a single window of an app, although it is moved to the right workspace, the order of the windows (both of an app and of different apps) is not preserved.
> 
> This was never implemented. But I thought about it and might do it at a later time.

What part of my use case was never implemented?
- move apps to the right workspace:
   - I think this is implemented;
- preserve order of the windows of different apps within a workspace;
- preserve order of the windows of a single apps within a workspace:
   - this is not yet implemented;
   - I see no way to identify a window across reboots if an app has multiple windows.

**Reply by Nama** [reply]
- [x] move apps to the right workspace
- [x] move all windows of a single app to the right workspaces (kinda, title needs to match)
- [ ] preserve order of the windows of different apps within a workspace
- [ ] preserve order of the windows of a single apps within a workspace (could be possible with title matching)

The remaining two should be possible to implement, but only with title matching for different windows of the same app.

**Reply by tukusejssirs** [reply]
> * [x]  move all windows of a single app to the right workspaces (kinda, title needs to match)

I don’t that for moving all windows of a single app to the right workspace needs to match the window title. All windows of a single app share the same `app_id`/`class`/`name` (unless the app dev forgot to fill it out, like Vorta packaged in Arch Linux AUR, currently has `app_id` set to `python3`). IMHO `app_id` should be considered a unique app ID, although I have no proof for that.

**Reply by Nama** [reply]
> In Sway config:

This should not interfere with `sway-workspaces`, since this is only applied while the window is created. But you run `sway-workspaces` after all your windows are created.
The same as, moving the window to another workspace via hotkey, even tho, it's assigned to a workspace.
